node/network/bridge/src/lib.rs

```diff
@@ -141,355 +138,566 @@ struct PeerData {
 
-			futures::select! {
-				subsystem_msg = subsystem_next => Action::from(subsystem_msg),
-				net_event = net_event_next => Action::from(net_event),
-				req_res_event = req_res_event_next  => Action::from(req_res_event),
-			}
-		};
-
-		// Used for logging purposes.
-		let before_action_process = Instant::now();
-
-		match action {
-			Action::Nop => {}
-			Action::Abort(reason) => match reason {
-				AbortReason::SubsystemError(err) => {
-					tracing::warn!(
-						target: LOG_TARGET,
-						err = ?err,
-						"Shutting down Network Bridge due to error"
-					);
-					return Err(SubsystemError::Context(format!(
-						"Received SubsystemError from overseer: {:?}",
-						err
-					)));
-				}
-				AbortReason::EventStreamConcluded => {
-					tracing::info!(
+	loop {
+		futures::select! {
+			msg = ctx.recv().fuse() => match msg {
+				Ok(FromOverseer::Signal(OverseerSignal::ActiveLeaves(active_leaves))) => {
+					let ActiveLeavesUpdate { activated, deactivated } = active_leaves;
+					tracing::trace!(
 						target: LOG_TARGET,
-						"Shutting down Network Bridge: underlying request stream concluded"
+						action = "ActiveLeaves",
+						num_activated = %activated.len(),
+						num_deactivated = %deactivated.len(),
 					);
-					return Err(SubsystemError::Context(
-						"Incoming network event stream concluded.".to_string(),
-					));
+
+					for activated in activated {
+						let pos = live_heads
+							.binary_search_by(|probe| probe.number.cmp(&activated.number).reverse())
+							.unwrap_or_else(|i| i);
+
+						live_heads.insert(pos, activated);
+					}
+					live_heads.retain(|h| !deactivated.contains(&h.hash));
+
+					update_our_view(
+						&mut network_service,
+						&mut ctx,
+						&live_heads,
+						&shared,
+						finalized_number,
+					).await?;
 				}
-				AbortReason::RequestStreamConcluded => {
-					tracing::info!(
+				Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(_hash, number))) => {
+					tracing::trace!(
 						target: LOG_TARGET,
-						"Shutting down Network Bridge: underlying request stream concluded"
+						action = "BlockFinalized"
 					);
-					return Err(SubsystemError::Context(
-						"Incoming network request stream concluded".to_string(),
-					));
-				}
-				AbortReason::OverseerConcluded => return Ok(()),
-			}
 
-			Action::SendValidationMessages(msgs) => {
-				tracing::trace!(
-					target: LOG_TARGET,
-					action = "SendValidationMessages",
-					num_messages = %msgs.len(),
-				);
+					debug_assert!(finalized_number < number);
 
-				for (peers, msg) in msgs {
-					send_message(
-							&mut bridge.network_service,
+					// we don't send the view updates here, but delay them until the next `ActiveLeaves`
+					// otherwise it might break assumptions of some of the subsystems
+					// that we never send the same `ActiveLeavesUpdate`
+					// this is fine, we will get `ActiveLeaves` on block finalization anyway
```

Comment:
this is no longer true I believe, which is why we're using `BlockFinalized` signal in approval-distibution instead of `OurViewChange`
```suggestion
```

Reply:
Hm, understood. I just moved this comment from the previous code.